src/routeGeneration/templateHelpers.ts

```diff
@@ -362,6 +366,155 @@ export class ValidationService {
+        );
+        cleanValues = {
+          ...cleanValues,
+          ...cleanValue,
+        };
+        subFieldErrors.push(subFieldError);
+      });
+
+    const filtered = subFieldErrors.filter(subFieldError => Object.keys(subFieldError).length !== 0);
+
+    if (filtered.length > 0) {
+      fieldErrors[parent + name] = {
+        message: `Could not match the intersection against every type. Issues: ${JSON.stringify(filtered)}`,
+        value,
+      };
+      return;
+    }
+
+    if (this.resolveAdditionalPropSetting(swaggerConfig) === 'silently-remove-extras') {
+      return cleanValues;
+    }
+
+    // Only Model definitions make sense here right now
+    const refNames = subSchemas.filter(subschema => subschema.ref).map(subschema => subschema.ref) as string[];
+    if (!refNames.every(refName => this.models[refName])) {
+      return value;
+    }
+
+    const reportedExcess = new Set(refNames.map(refName => this.models[refName]).reduce((acc, subSchema) => {
+      return [...acc, ...this.getExcessPropertiesFor(subSchema, Object.keys(value), swaggerConfig)];
+    }, []));
+
+    if (reportedExcess.size === 0) {
+      return value;
+    }
+
+    const allowedProperties = new Set(refNames.map(refName => this.models[refName]).reduce((acc, subSchema) => {
+      return [...acc, ...this.getPropertiesFor(subSchema)];
+    }, []));
+
+    const actualExcess = [...reportedExcess].filter(property => !allowedProperties.has(property));
+
+    if (actualExcess.length > 0) {
+      fieldErrors[parent + name] = {
+        message: `The following properties are not allowed by any part of the intersection: ${actualExcess}`,
+        value,
+      };
+    }
+
+    return value;
+  }
+
+  private resolveAdditionalPropSetting(swaggerConfig: SwaggerConfigRelatedToRoutes): AdditionalPropSetting {
+    if (!swaggerConfig.noImplicitAdditionalProperties) {
+      return 'ignore';
+    } else if (swaggerConfig.noImplicitAdditionalProperties === 'throw-on-extras' || swaggerConfig.noImplicitAdditionalProperties === true) {
+      return 'throw-on-extras';
+    } else if (swaggerConfig.noImplicitAdditionalProperties === 'silently-remove-extras') {
+      return 'silently-remove-extras';
+    } else {
+      return assertNever(swaggerConfig.noImplicitAdditionalProperties);
+    }
+  }
+
+  private getPropertiesFor(modelDefinition: TsoaRoute.ModelSchema) {
+    return new Set(Object.keys((modelDefinition && modelDefinition.properties) || {}));
+  }
+
+  private getExcessPropertiesFor(
+    modelDefinition: TsoaRoute.ModelSchema,
+    properties: string[],
+    config: SwaggerConfigRelatedToRoutes,
+  ): string[] {
+    if (!modelDefinition || !modelDefinition.properties) {
+      return properties;
+    }
+
+    const modelProperties = new Set(Object.keys(modelDefinition.properties));
+
+    if (modelDefinition.additionalProperties) {
+      return [];
+    } else if (!config.noImplicitAdditionalProperties) {
```

Comment:
It took me a while to figure out how it was possible that the unit test would even pass, and it turns out that it's because of this line. Please change this to exhaustively check each of the members of the union so that the typescript compiler will inform us if a new member is added to the `noImplicitAdditionalProperties` union and if we haven't considered it here. And if that seems far fetched, consider that you asked for their to be an `"ignore"` option which would you would not want to return `[]` for.

So the code should look like this:
```ts
    if (modelDefinition.additionalProperties) {
      return [];
    } else {
      if (
        config.noImplicitAdditionalProperties === false ||
        config.noImplicitAdditionalProperties === undefined
      ) {
        return [];
      } else if (
        config.noImplicitAdditionalProperties === 'throw-on-extras' ||
        config.noImplicitAdditionalProperties === true ||
        config.noImplicitAdditionalProperties === 'silently-remove-extras'
      ) {
        return [...properties].filter(property => !modelProperties.has(property));
      } else {
        assertNever(config.noImplicitAdditionalProperties);
      }
    }
```

And sorry, I said that I wouldn't make any more corrections. But if you are going to get in the code to fix this other (admittedly minor) request https://github.com/lukeautry/tsoa/pull/400/files#r316668417, then maybe you can make this exhaustiveness checking change too.

Reply:
Ok, so there are a few notes:

a) `!config.noImplicitAdditionalProperties` is equal to 
```
config.noImplicitAdditionalProperties === false ||
config.noImplicitAdditionalProperties === undefined
```

b) the `else if` should only catch `throw` and `true`, I'll push a commit to reuse the `resolveAdditionalProps` helper I added.

c) the `else if` can't be `silently remove` because if this were true the validation would've already returned [here](https://github.com/lukeautry/tsoa/pull/400/files#diff-054445345683aa77f57a22be4b3d8b28R400) or [here](https://github.com/lukeautry/tsoa/pull/400/files#diff-054445345683aa77f57a22be4b3d8b28R449).
However, factoring in b) since now the `resolveAdditionalPropSetting` helper is [used](https://github.com/lukeautry/tsoa/pull/400/commits/51d50e0917859b48be5cf2ee75617709fd8f7ffa), this branch can only be `silently-remove` or `throw` (due to exhaustiveness of `resolveAdditionalPropSetting`), which is equivalent to the code you suggested.

Reply:
> !config.noImplicitAdditionalProperties is equal to
I know that. It's just not very "scan-able." Your newest push is much easier to read and to discern the meaning. Thank you.